### PR TITLE
test(title-filter): add a real-title recall regression corpus

### PR DESCRIPTION
**Pull request by IaroslavMazur**
Takes up the standing invitation in the closing comment of #3104.

## Problem

A change to keyword matching can stop matching real roles with no visible symptom. The scan summary reports one lower number, which reads exactly like a quiet week on the boards. #2544 made the same observation about the single "filtered by title" counter: a well-tuned filter and a leaking one look identical from the outside.

#2970 settled the anchoring default on exactly this kind of evidence. Measuring the proposed anchor-by-default against a real corpus showed it dropped `Engineering Lead - Crypto and DeFi`, and that one recall loss is what decided the design. The corpus was in my private `scan-history.tsv`, so:

1. **Nobody else could check it**: the measurement that settled a repo-wide default was unverifiable by the people it binds.
2. **The next proposal has to rebuild one**: from scratch, in someone else's private data, with no guarantee the two corpora are comparable.
3. **Nothing holds the guarantee afterwards**: the decision is recorded in issue comments, not in anything that fails.

## Change

- **`tests/fixtures/title-recall-corpus.json`**: 675 distinct titles observed on public Greenhouse, Lever, Ashby, Workday and iCIMS boards between 2026-04-20 and 2026-08-22, each with the verdict the current matcher returns. Titles only, i.e., no company, URL, board, or date.
- **`tests/title-recall-corpus.test.mjs`**: replays the corpus, names the drifted titles instead of counting them, and pins the ten specific cases #3103 and #2970 discussed.

Two deliberate details:

1. **The fixture ships its own 11-keyword filter** rather than copying a real `portals.yml`. It holds exactly the keywords whose behaviour those issues discussed, so the corpus exercises the documented edge cases without depending on a user config that can change underneath the test.
2. **A second assertion guards the fixture itself.** An emptied corpus, or a filter reduced to a no-op, would satisfy the main assertion forever while guarding nothing. It pins both edges: 386 titles match, 289 do not.

A diff in the fixture is a decision, not automatically a bug. The intended workflow is to regenerate it in the same commit that changes the matcher, and say in the message which titles moved and why. Failure output names the first eight, split into lost and gained, because "12 titles changed" sends you to a diff whereas `Solana Program Engineer no longer matches` sends you to the cause.

## Measured

Seen RED under three deliberate regressions:

| regression | failures |
|---|---|
| anchor everything by default (the #3104 design) | 5 |
| `word:` prefix ignored, treated as a literal | 3 |
| AND-groups collapsed to plain substring | 3 |

`node test-all.mjs`: 5197 passed, 0 failed, 3 warnings (all environmental - no user data in a clean checkout, no Go compiler for the dashboard build).

One thing the corpus surfaced while I was building it, as a sample of what it is for: `Astar` matches inside `NASTAR`, so a ski resort's `NASTAR/Events Crew - Winter 26-27` came through a reverse-ATS sweep as a smart-contract role. `word:Astar` closes it and keeps `Astar Network Rust Engineer`. That case is pinned in the test.

## Alternatives considered

- **A curated set of ~40 edge cases instead of the full corpus.** Smaller and easier to read, but it only guards the bleeds we already know about. The point of a recall corpus is the 289 non-matching titles nobody thought to write down.
- **Generating expected values from the matcher at test time** rather than recording them. That is what `tests/fixtures/company-key-corpus.json` does, and it is right there because the core is the source of truth for a mirror. Here it would assert nothing at all: the matcher would always agree with itself.

Happy to trim the corpus or split the pinned cases into their own file if either reads as too much for one fixture.

Does the regenerate-in-the-same-commit workflow match how you would want a matcher change handled?

## Summary by CodeRabbit

* **Tests**
  * Added a title-matching regression corpus covering backend, protocol, Rust, Solana, smart-contract, DeFi, frontend, recruiting, internship, and false-positive cases.
  * Added validation for expected matches, non-matches, synthetic examples, and documented historical cases.
  * Improved detection of matching drift, including unexpectedly lost or gained results.
  * Added safeguards to reject incomplete or degenerate test data.

### CODE REVIEW

File: tests/title-recall-corpus.test.mjs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  ['word:Solana', 'Genomic Breeder (Solanaceae)', false, '#2970: word: closes it'],
+  ['word:Solana', 'Senior Solana Engineer', true, 'word: keeps the real match'],
+  ['Rust Engineer', 'Sr. Zero Trust Engineer III (6794)', true, '#3103: "Trust Engineer" inside'],
+  ['word:DeFi + Engineer', 'Engineering Lead - Crypto and DeFi', true, '#2970: the recall case that settled the default'],
+  ['word:DeFi + Engineer', 'Senior Systems Engineer, Product Definition', false, 'word: composes inside an AND-group'],
+  ['DeFi + Engineer', 'Senior Systems Engineer, Product Definition', true, 'control: unprefixed AND-group still bleeds'],
+  ['word:Astar', 'NASTAR/Events Crew - Winter 26-27', false, 'observed 2026-08-22: Astar inside NASTAR'],
+  ['word:Astar', 'Astar Network Rust Engineer', true, 'and the real one survives'],
+  ['C++', 'Senior C++ Engineer', true, 'non-word edges keep substring matching — \\b could never match here'],
+];
```
[the remark this answers is not in the document]
Comment (IaroslavMazur) [reply tests/title-recall-corpus.test.mjs]: > Both cases stay valid as synthetic controls. Correct the comment so a reader does not search the fixture for them.

Correct, and thanks for the catch, @coderabbitai! 🙏

I took the fix one step further than the suggested diff, though. Correcting the sentence fixes this instance; it does not stop the next case added to `CASES` from reintroducing the same mismatch between what the comment claims and what the fixture holds. So the two are now marked `SYNTHETIC` in the array, and a fourth assertion checks the marking against the corpus in both directions: 1) a case marked synthetic that turns out to be observed fails, and 2) an unmarked case that is not in the corpus fails.

Seen RED by dropping the marking from one entry: `unmarked but NOT in the corpus: Astar Network Rust Engineer`.

Pushed as `d9eee25f`.

**Review by Scott-Emberson — APPROVED** [review]
Reviewed as code owner of `tests/`. Scope is the two new files, both under `tests/`, so this one is entirely mine.

This is a real recall guard, not a vacuous one. It imports `buildTitleFilter` from the same shared matcher `scan.mjs` uses, so there is no reimplementation. The first assertion fails and names any title the current matcher scores differently from its recorded verdict, which is exactly the recall regression a narrowed filter would cause. The anti-degeneracy assertion is the part that makes it honest: it fails if the corpus matches nothing or everything, closing the silently-emptied-fixture hole, and the corpus is not degenerate at 386 match against 289 no-match. The corpus is realistic ATS titles rather than strings built to pass, and it pins the #3103 and #2970 bleeds by name (the Solanaceae substring, "Trust Engineer" inside "Zero Trust", the DeFi recall case).

Hermetic, fixture read from disk. Green on CI. Approving from my side.

**Comment by santifer** [comment]
Another solid one, @IaroslavMazur, and this one closes a loop I care about: it is invitation (1) from the #3104 close, delivered exactly as scoped. The corpus that decided the default now lives where anyone can check it, instead of inside a private scan-history.tsv nobody else could open. Two things I verified rather than took on trust: the fixture carries titles only (no emails, no URLs, no employers), and the anti-degeneration assertion means the corpus cannot be silently emptied into a rubber stamp. You also shipped two things nobody asked for: that guard, and the ten pinned cases from #3103 and #2970. One small thing for a future pass: twelve of the entries are scan-history annotation rows rather than titles (all inert with matches:false), so the provenance line saying 675 distinct titles is a hair off. Merged! 🚀
